### PR TITLE
error state implemented

### PR DESCRIPTION
**Pull request by Vuk-Mijatovic**

### CODE REVIEW

File: app/src/main/java/com/example/booklisting/MainActivity.java [hunk 1]
```diff
@@ -130,38 +125,42 @@ public void onLoadMore() {
     }
 
     @Override
-    public void onLoadFinished(@NonNull Loader<List<Book>> loader, List<Book> list) {
+    public void onLoadFinished(@NonNull Loader<BookLoaderResult> loader, BookLoaderResult loaderResult) {
         if ((adapter == null) || (adapter.getItemCount() == 0)) {
-            if (exceptionThrown) {
+            if (loaderResult.getException() != null) {
+                Log.i("Exception is", String.valueOf(loaderResult.getException()));
```
Comment (nemanja-kovacevic) [inline app/src/main/java/com/example/booklisting/MainActivity.java]: Log code should be used only for local test, it should not be committed or merged to master

File: app/src/main/java/com/example/booklisting/MainActivity.java [hunk 2]
```diff
@@ -130,38 +125,42 @@ public void onLoadMore() {
...
                 progressBar.setVisibility(View.GONE);
                 emptyView.setVisibility(View.GONE);
                 showAlert();
             } else {
 
-                if (list.isEmpty()) {
+                if (loaderResult.getResult().isEmpty()) {
```
Comment (nemanja-kovacevic) [inline app/src/main/java/com/example/booklisting/MainActivity.java]: Are you sure there is not possibility of NPE here? (NullPointerException). If there is a error, result array inside loaderResult could be null if I'm not mistaken?

File: app/src/main/java/com/example/booklisting/MainActivity.java [hunk 3]
```diff
@@ -130,38 +125,42 @@ public void onLoadMore() {
...
                     bookList.setAdapter(adapter);
                 }
             }
         } else {
-            if (exceptionThrown) {
+            if (loaderResult.getResult().isEmpty()) {
                 emptyView.setVisibility(View.GONE);
                 progressBar.setVisibility(View.GONE);
-                showAlert();
+                if (books.get(books.size() - 1) == null) {
```
Comment (nemanja-kovacevic) [inline app/src/main/java/com/example/booklisting/MainActivity.java:157]: I'm not clear what is this check actually checking?

**Review by nemanja-kovacevic — COMMENTED** [review]
In total, this is a good solution to the problem "how can I report back if error happened during loading"

File: app/src/main/java/com/example/booklisting/MainActivity.java (hunk 1 above)
Comment (Vuk-Mijatovic) [reply app/src/main/java/com/example/booklisting/MainActivity.java]: Corrected.

File: app/src/main/java/com/example/booklisting/MainActivity.java (hunk 3 above)
Comment (Vuk-Mijatovic) [reply app/src/main/java/com/example/booklisting/MainActivity.java:157]: It was naive solution for problem that i had with deleting last entry, when new set of data is loaded. I think that that bug was resolved whit proper implementation of onScrolllistener. Anyway it works fine now, so I removed that check.

File: app/src/main/java/com/example/booklisting/MainActivity.java (hunk 2 above)
Comment (Vuk-Mijatovic) [reply app/src/main/java/com/example/booklisting/MainActivity.java]: You`re right, I missed that. I added another check at the beginning of onLoadFinished() method.

**Review by nemanja-kovacevic — APPROVED** [review]

**Pull request merged**